### PR TITLE
libllvm: update from 10.0.1 to 11.0.0 and merge lldb

### PR DESCRIPTION
**Pull request by buttaface**
Use the unified source, as we're already downloading most of it separately anyway. Once the build artifacts are up, I'll need @its-pointless, @Grimler91, and others to test on arches other than aarch64 and let me know about any regressions.

### CODE REVIEW

File: packages/libllvm/lldb.subpackage.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+TERMUX_SUBPKG_INCLUDE="
+bin/lldb*
+include/lldb/
+lib/liblldb.so
+share/man/man1/lldb.1.gz
+"
+TERMUX_SUBPKG_DESCRIPTION="LLVM-based debugger"
```
Comment (buttaface) [inline packages/libllvm/lldb.subpackage.sh:7]: The only differences I could find from building this as a subpackage are that subpackages can't have their own homepage and they don't have a `LICENSE` file.

File: packages/libllvm/build.sh [hunk 2]
```diff
@@ -64,64 +54,59 @@ TERMUX_PKG_HAS_DEBUG=false
...
-		mkdir projects/openmp/runtime/src/android
-		cp $TERMUX_PKG_BUILDER_DIR/nl_types.h projects/openmp/runtime/src/android
-		cp $TERMUX_PKG_BUILDER_DIR/nltypes_stubs.cpp projects/openmp/runtime/src/android
+		mkdir openmp/runtime/src/android
+		cp $TERMUX_PKG_BUILDER_DIR/nl_types.h openmp/runtime/src/android
+		cp $TERMUX_PKG_BUILDER_DIR/nltypes_stubs.cpp openmp/runtime/src/android
 	fi
 
-	export LLVM_DEFAULT_TARGET_TRIPLE=$TERMUX_HOST_PLATFORM
+	export LLVM_DEFAULT_TARGET_TRIPLE=$CCTERMUX_HOST_PLATFORM
```
Comment (buttaface) [inline packages/libllvm/build.sh:72]: Now including the Android API level in the default target triple, which was done to some extent already by the `Triple.h` patch, that's now removed as not needed after this change.

**Comment by buttaface** [comment]
I can't unzip the zip file with the CI-built packages: corrupted download on my part or others have the same problem?

**Comment by Grimler91** [comment]
> I can't unzip the zip file with the CI-built packages: corrupted download on my part or others have the same problem?

No problems for me when I tried it just now

**Comment by buttaface** [comment]
What's weird is that it's the right size, `ls -l` says 395788280, and sha256 of 17a2ec1a956a49b66447fc7ccb6eb35e7b0980a791ec5b11546dc15bbfab403f, same for you? The `unzip` command in Termux says it can't open it. Anyway, I kicked off a build for AArch64, will just test that.

**Comment by Grimler91** [comment]
Size 396236964 and sha256 fa57a1f68a824561e95bbf82428bc944013445b4d8457fd7a043e70a7bbffc39 so seem to be a corrupt download you have there (does `file termux-packages.zip` identify it as a zip?)

**Comment by buttaface** [comment]
```
> file termux-packages.zip
termux-packages.zip: Zip archive data, at least v2.0 to extract
```
That is the worst scenario, almost downloads the whole thing but fails at the end without error: I really hate the Chrome downloader. Anyway, thanks for confirming that was it, now just let me know how it works on ARM. 😄

**Comment by buttaface** [comment]
@its-pointless, the rust package will need to be updated after this: is a simple version bump in there enough or will it take more work?

**Comment by Grimler91** [comment]
> Anyway, thanks for confirming that was it, now just let me know how it works on ARM

Yes, my arm tablet is broken at the moment, but can test on an old galaxy s4 (with LOS 17.1) tonight or tomorrow!

File: packages/libllvm/clang-lib-Driver-ToolChains-Linux.cpp.patch [hunk 3]
```diff
@@ -28,7 +28,7 @@
  
    addPathIfExists(D, SysRoot + "/lib", Paths);
 -  addPathIfExists(D, SysRoot + "/usr/lib", Paths);
-+  bool nativeBuild = MultiarchTriple == llvm::sys::getDefaultTargetTriple();
++  bool nativeBuild = MultiarchTriple == getMultiarchTriple(D, llvm::Triple(llvm::sys::getDefaultTargetTriple()), SysRoot);
```
Comment (buttaface) [inline packages/libllvm/clang-lib-Driver-ToolChains-Linux.cpp.patch:31]: Sigh, this line still isn't right, been struggling with getting it to work and submitted without making sure this version worked. I will update this pull once I'm sure I have it.

**Comment by its-pointless** [comment]
> @its-pointless, the rust package will need to be updated after this: is a simple version bump in there enough or will it take more work?

last time i tried updating rust it reached disk limit on github. That was unstable version though.

**Comment by buttaface** [comment]
We could clear out more space on the Ubuntu VM used by the github CI. I've been looking into what's in there lately and [it comes with a ton of tools we don't use, including a copy of the latest Android NDK](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-README.md). Since we use our own Docker setup inside of that Ubuntu VM, we could clear a lot more of those unused tools out of the VM.

**Comment by buttaface** [comment]
Alright, seems to work on aarch64 now, let me know what you see when using these LLVM packages in your projects.

**Comment by its-pointless** [comment]
I just remembered that wasm32-unknown-unknown rust doesn't build properly unless we use llvm-11 to build rust.  This combined with disk limitations which i tried to fix but can't at least with unstable. It's likely that stable rust doesn't use as much disk space when building as unstable branch of rust. 

I will test it at some point in the next day or so for sure.

**Comment by Grimler91** [comment]
Arm build seem to work. I have tried building a couple of different project and had no problems

**Comment by buttaface** [comment]
Great, we'll merge once @its-pointless confirms that the rust package is ready for this change. I will need to bump the Swift package, which depends on this too, so we can push those in separate commits right after this one.

**Comment by its-pointless** [comment]
Yeah i have rust building 1.48 testing how much disk space it uses. Its likely we have to change the docker image to use 20.10 and edit the setup-ubuntu.sh to get llvm-11 instead of llvm-10. The wasm32 code from llvm-10 and 11 don't work with each other.

**Comment by buttaface** [comment]
Rebased, moved another OpenMP header into the clang package, and moved all executables out of libllvm into the right packages.

> edit the setup-ubuntu.sh to get llvm-11 instead of llvm-10. The wasm32 code from llvm-10 and 11 don't work with each other

The native Ubuntu package is only used to compile this LLVM 11 and the Rust package, so I think clang 10 should be fine for that. I don't care if those Ubuntu changes are made in later pulls, but I don't think they're necessary for the Rust update to use this LLVM 11, so let's keep this simple.

@xeffyr, what do you think?

**Comment by its-pointless** [comment]
"but I don't think they're necessary for the Rust update to use this LLVM 11, so let's keep this simple."

rust builds a std library for each target. So it is also built for the wasm32-unknown-unknown.  When llvm 10 outputs wasm32 it isn't compatible with the llvm wasm32 output and std image. When you try to compile rust and target wasm32 on device it will fail.

As it stand right now i don't think its possible to build rust on github using github workflows because there is less disk space than there used to be. Before they weren't enforcing the cap on disk usage and now its 12 gig.

**Comment by buttaface** [comment]
> When llvm 10 outputs wasm32 it isn't compatible with the llvm wasm32 output and std image.

OK, you are generating the Rust wasm stdlib in Ubuntu x86_64, but that would be with a x86_64 prebuilt rustc that you downloaded and is built against LLVM 11, so it should be fine? You're saying that generating the wasm stdlib on Ubuntu x86_64 somehow uses the clang/LLVM 10 prebuilt packages in a way that corrupts building that wasm stdlib? Or that you're building rustc from scratch for x86_64 also and that needs to be linked against LLVM 11, before that newly built x86_64 rustc is used to build the wasm stdlib?

> Before they weren't enforcing the cap on disk usage and now its 12 gig

What makes you think there is a cap? I don't see it mentioned anywhere in their CI docs. Rather, they provide an Ubuntu VM within which we run our Docker container and their VM only has something like 15-20 GBs of free space, which we add to [by removing unused packages in that VM](https://github.com/termux/termux-packages/blob/master/.github/workflows/packages.yml#L27). We could remove a lot more of the unused dev tools I linked to above and have a lot more free space, I will open a separate pull for that.

**Comment by landfillbaby** [comment]
just curious but is anything still holding up the merge at this point?

**Comment by buttaface** [comment]
[Maybe](https://github.com/termux/termux-packages/pull/6162#issuecomment-750659207).

**Pull request merged**